# 📅 Day 5 of 100 — Data Types & Missing Values

📂 **GitHub Repo:** [100-Days-of-Data-Science / Day 005](https://github.com/imshoaibaslam/100-Days-of-Data-Science/tree/main/Day-06-Datascience)

---

## 🗂️ Agenda
1. Understanding Data Types with `dtype` and `dtypes`
2. Converting Data Types with `astype()`
3. Finding Missing Values with `pd.isnull()` and `pd.notnull()`
4. Filling Missing Values with `fillna()`
5. Replacing Values with `replace()`

---

## 🎯 Deliverables
- [ ] Check the dtype of individual columns and the full DataFrame
- [ ] Convert a column from one dtype to another using `astype()`
- [ ] Find all rows where a column contains NaN
- [ ] Fill missing values using `fillna()`
- [ ] Replace specific values using `replace()`

---

## 📚 Learning Topics

| # | Topic | Description |
|---|-------|-------------|
| 1 | `dtype` | Check the data type of a single column |
| 2 | `dtypes` | Check data types of all columns at once |
| 3 | `astype()` | Convert a column from one type to another |
| 4 | `pd.isnull()` | Find rows with missing (NaN) values |
| 5 | `fillna()` | Fill missing values with a default or strategy |
| 6 | `replace()` | Replace specific existing values in a column |





## 🔧 Imports



In [1]:
import pandas as pd
import numpy as np

---

## Step 0 — Load the Dataset

In [2]:
# Load the Wine Reviews dataset
reviews = pd.read_csv(r"C:\Users\aslam\Downloads\winemag-data-130k-v2.csv\winemag-data-130k-v2.csv", index_col=0)
reviews.head(3)

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,87,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm


---

## Step 1 — Understanding Data Types

### What is a `dtype`?

Every column in a DataFrame has a **data type** — it tells pandas how to store and process that column internally.

| dtype | Meaning |
|-------|---------|
| `float64` | Decimal numbers |
| `int64` | Whole numbers |
| `object` | Text / strings |
| `bool` | True / False |

⚠️ **Important:** String columns do not get a `string` type in pandas.
They are stored as `object` — keep this in mind when inspecting your data.

#### `dtype` — Check a Single Column

In [3]:
# What data type is the price column stored as?
reviews.price.dtype

dtype('float64')

#### `dtypes` — Check All Columns at Once

In [4]:
# Get the data type of every column in one shot
reviews.dtypes

country                   object
description               object
designation               object
points                     int64
price                    float64
province                  object
region_1                  object
region_2                  object
taster_name               object
taster_twitter_handle     object
title                     object
variety                   object
winery                    object
dtype: object

#### Index `dtype`

In [5]:
# The index also has its own dtype
reviews.index.dtype

dtype('int64')

---

## Step 2 — Converting Data Types with `astype()`

### What is `astype()`?

Sometimes the data type pandas assigned is not the one you need for your analysis.
`astype()` lets you **convert a column from one type to another**.

Example use case:
- `points` is stored as `int64` but your model expects `float64`
- A `0/1` column stored as `int64` should be `bool`

⚠️ The conversion only makes sense if the values are compatible.
Converting a text column to `int64` will throw an error.

In [6]:
# Convert points from int64 to float64
reviews.points.astype('float64')

0         87.0
1         87.0
2         87.0
3         87.0
4         87.0
          ... 
129966    90.0
129967    90.0
129968    90.0
129969    90.0
129970    90.0
Name: points, Length: 129971, dtype: float64

**What just happened?**

- `points` was originally stored as whole numbers (`int64`)
- `astype('float64')` converted every value to a decimal
- 87 → 87.0 — same value, different storage type
- Original column is untouched — `astype()` returns a new Series

---

## Step 3 — Finding Missing Values

### What is `NaN`?

Missing values in pandas are stored as `NaN` — short for **Not a Number**.

```
Real value  →  87, 15.0, "Italy"
Missing     →  NaN
```

⚠️ NaN values are always `float64` dtype — even if the column contains integers.
This is why a column of whole numbers sometimes shows as `float64` — it has at least one missing value.

### `pd.isnull()` vs `pd.notnull()`

| Function | Returns |
|----------|--------|
| `pd.isnull()` | `True` where value is NaN |
| `pd.notnull()` | `True` where value is NOT NaN |

#### Find All Rows Where Country is Missing

In [7]:
# Show all rows where country has no value
reviews[pd.isnull(reviews.country)]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
913,NaN,"Amber in color, this wine has aromas of peach ...",Asureti Valley,87,30.0,NaN,NaN,NaN,Mike DeSimone,@worldwineguys,Gotsa Family Wines 2014 Asureti Valley Chinuri,Chinuri,Gotsa Family Wines
3131,NaN,"Soft, fruity and juicy, this is a pleasant, si...",Partager,83,NaN,NaN,NaN,NaN,Roger Voss,@vossroger,Barton & Guestier NV Partager Red,Red Blend,Barton & Guestier
4243,NaN,"Violet-red in color, this semisweet wine has a...",Red Naturally Semi-Sweet,88,18.0,NaN,NaN,NaN,Mike DeSimone,@worldwineguys,Kakhetia Traditional Winemaking 2012 Red Natur...,Ojaleshi,Kakhetia Traditional Winemaking
9509,NaN,This mouthwatering blend starts with a nose of...,Theopetra Malagouzia-Assyrtiko,92,28.0,NaN,NaN,NaN,Susan Kostrzewa,@suskostrzewa,Tsililis 2015 Theopetra Malagouzia-Assyrtiko W...,White Blend,Tsililis
9750,NaN,This orange-style wine has a cloudy yellow-gol...,Orange Nikolaevo Vineyard,89,28.0,NaN,NaN,NaN,Jeff Jenssen,@worldwineguys,Ross-idi 2015 Orange Nikolaevo Vineyard Chardo...,Chardonnay,Ross-idi
...,...,...,...,...,...,...,...,...,...,...,...,...,...
124176,NaN,This Swiss red blend is composed of four varie...,Les Romaines,90,30.0,NaN,NaN,NaN,Jeff Jenssen,@worldwineguys,Les Frères Dutruy 2014 Les Romaines Red,Red Blend,Les Frères Dutruy
129407,NaN,Dry spicy aromas of dusty plum and tomato add ...,Reserve,89,22.0,NaN,NaN,NaN,Michael Schachner,@wineschach,El Capricho 2015 Reserve Cabernet Sauvignon,Cabernet Sauvignon,El Capricho
129408,NaN,El Capricho is one of Uruguay's more consisten...,Reserve,89,22.0,NaN,NaN,NaN,Michael Schachner,@wineschach,El Capricho 2015 Reserve Tempranillo,Tempranillo,El Capricho
129590,NaN,"A blend of 60% Syrah, 30% Cabernet Sauvignon a...",Shah,90,30.0,NaN,NaN,NaN,Mike DeSimone,@worldwineguys,Büyülübağ 2012 Shah Red,Red Blend,Büyülübağ


#### Count Missing Values Per Column

In [8]:
# How many missing values does each column have?
reviews.isnull().sum()

country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64

---

## Step 4 — Filling Missing Values with `fillna()`

### What is `fillna()`?

Instead of dropping rows with missing values, we can **fill them** with something meaningful.

Two common strategies:
- Fill with a **constant** — replace NaN with a default label like `"Unknown"`
- Fill with **backfill** — use the next non-null value that appears after the missing one

#### Fill Missing Values with a Constant

In [9]:
# Replace every NaN in region_2 with the string "Unknown"
reviews.region_2.fillna("Unknown")

0                   Unknown
1                   Unknown
2         Willamette Valley
3                   Unknown
4         Willamette Valley
                ...        
129966              Unknown
129967         Oregon Other
129968              Unknown
129969              Unknown
129970              Unknown
Name: region_2, Length: 129971, dtype: object

**What just happened?**

- Every row where `region_2` was `NaN` now shows `"Unknown"`
- Rows that already had a value are untouched
- Original data is not modified — `fillna()` returns a new Series

#### Fill Missing Values with the Column Mean

In [10]:
# Replace missing prices with the average price
reviews.price.fillna(reviews.price.mean())

0         35.363389
1         15.000000
2         14.000000
3         13.000000
4         65.000000
            ...    
129966    28.000000
129967    75.000000
129968    30.000000
129969    32.000000
129970    21.000000
Name: price, Length: 129971, dtype: float64

---

## Step 5 — Replacing Values with `replace()`

### What is `replace()`?

`replace()` is used when you want to **swap out a specific existing value** — not just fill NaN.

Real-world use case:
- A reviewer changed their Twitter handle
- A category label was renamed
- A sentinel value like `"N/A"` or `"Invalid"` needs to be replaced

In [11]:
# Reviewer Kerin O'Keefe changed her handle from @kerinokeefe to @kerino
# Update the entire column to reflect this
reviews.taster_twitter_handle.replace("@kerinokeefe", "@kerino")

0             @kerino
1          @vossroger
2         @paulgwine 
3                 NaN
4         @paulgwine 
             ...     
129966            NaN
129967    @paulgwine 
129968     @vossroger
129969     @vossroger
129970     @vossroger
Name: taster_twitter_handle, Length: 129971, dtype: object

**What just happened?**

- Every occurrence of `@kerinokeefe` across all 129,971 rows was replaced with `@kerino`
- All other values remained unchanged
- Original column untouched — returns a new Series

#### `fillna()` vs `replace()` — Key Difference

| | `fillna()` | `replace()` |
|-|-----------|------------|
| Works on | NaN values only | Any existing value |
| Use when | Values are missing | Values exist but are wrong |

---

## 💡 When to Use What

| Situation | Use |
|-----------|-----|
| Check type of one column | `df['col'].dtype` |
| Check types of all columns | `df.dtypes` |
| Convert column to different type | `astype('float64')` |
| Find rows with missing values | `pd.isnull(df['col'])` |
| Count missing values per column | `df.isnull().sum()` |
| Fill NaN with a default value | `fillna('Unknown')` |
| Fill NaN with column average | `fillna(df['col'].mean())` |
| Replace a specific existing value | `replace('old', 'new')` |

---

## 🌍 Real World Usage

**`dtypes` + `astype()`**
The first check after loading any new dataset. Wrong data types cause silent bugs in machine learning pipelines — a price column stored as `object` instead of `float64` will break every numeric operation downstream.

**`isnull().sum()`**
Standard step in every Exploratory Data Analysis. Tells you exactly which columns have missing data and how much — critical before any modelling or aggregation.

**`fillna()`**
Used in data cleaning pipelines before training ML models. Models cannot handle NaN values — filling them with the mean, median, or a placeholder is a standard preprocessing step.

**`replace()`**
Used when datasets contain sentinel values like `"N/A"`, `"Unknown"`, `"Invalid"` — or when category labels change over time and historical data needs updating.

---

## 🔑 Key Takeaways

- Every column has a `dtype` — always check it after loading a new dataset
- String columns are stored as `object` — not `string`
- Missing values are stored as `NaN` and are always `float64` dtype
- `astype()` converts types but does not modify the original column
- `fillna()` fills NaN values — `replace()` swaps any existing value
- Neither `fillna()` nor `replace()` modifies original data — they return new copies
- Always run `df.isnull().sum()` as part of your first look at any new dataset

---

## 🏁 Conclusion

Today we covered **data types and missing values** — two of the most critical concepts in real-world data cleaning.

Before any analysis or modelling can happen, your data needs to be in the right format and free of unexpected gaps. Wrong dtypes cause silent failures. Unhandled NaN values break pipelines. These are not edge cases — they appear in every real dataset.

From Day 8 onwards, we start combining everything we have learned into complete end-to-end data cleaning workflows.

---

## 📝 Personal Notes & Observations

> Write anything you learned, got confused about, or want to remember here.

---

## ✅ Day 5 Complete — 95 days to go!

<p align="center">
  <a href="https://github.com/imshoaibaslam/100-Days-of-Data-Science/tree/main/Day-06-Datascience">📂 View Full Code on GitHub</a>
</p>